### ISSUE TITLE
MoE router replication is broken w/ TP

### ISSUE THREAD

**Issue by volcacius** [pytorch/torchtitan#2387]
### Bug description

The router replication is broken whenever TP is enabled in a MoE network, the weight diverges right after the first step among different TP ranks.
The issue is that the gradient coming from downstream the router is not being reduced. Changing outputs.to_local()
to outputs.to_local(grad_placements=(Partial(),)) in NoParallel fixes it. It's a similar type of bug as the one fixed here https://github.com/pytorch/torchtitan/commit/3ff755178cb65bd03d9c2766dba56ca45a04c8ef .

I can’t think of a worse class of bugs for MoEs to be honest, considering the router drives the majority of the weights. I’m curious if the recent proposal on DTensor erasure https://blog.ezyang.com/2026/02/dtensor-erasure/ would have caught it, or not really since this is mixed local/global SPMD scenario.

### Versions

Tested with torchtitan commit cf86edc977fb629fd8d93d1e0d1e2e8cb21a17d4 from a few days ago and looking at recent changes I don't see anything that could have fixed it.

**Comment by tianyu-l** [comment]
Definitely should bandaid such bugs and improve with strict tying later.
cc @acisseJZhong

**Comment by tianyu-l** [comment]
Can you give me the parallelism config you are using? I think I know one bug when TP > 1 and ETP=1 caused by this line missing a reduction in backward.
https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama4/infra/parallelize.py#L536
https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/expert_parallel.py#L285

**Comment by volcacius** [comment]
I just re-tested over a single 8 gpu node and it fails with any TP configuration, so:

TP=2, EP=1, ETP=1
```
CONFIG_FILE="./torchtitan/models/deepseek_v3/train_configs/debug_model.toml"   NGPU=8   ./run_train.sh   --parallelism.pipeline_parallel_degree 2   --parallelism.data_parallel_shard_degree 2   --parallelism.tensor_parallel_degree 2   --parallelism.expert_parallel_degree 1   --parallelism.expert_tensor_parallel_degree 1
```

TP=2, EP=2, ETP=1
```
CONFIG_FILE="./torchtitan/models/deepseek_v3/train_configs/debug_model.toml"   NGPU=8   ./run_train.sh   --parallelism.pipeline_parallel_degree 2   --parallelism.data_parallel_shard_degree 2   --parallelism.tensor_parallel_degree 2   --parallelism.expert_parallel_degree 2   --parallelism.expert_tensor_parallel_degree 1
```

TP=2,EP=2,ETP=2
```
CONFIG_FILE="./torchtitan/models/deepseek_v3/train_configs/debug_model.toml"   NGPU=8   ./run_train.sh   --parallelism.pipeline_parallel_degree 2   --parallelism.data_parallel_shard_degree 2   --parallelism.tensor_parallel_degree 2   --parallelism.expert_parallel_degree 2   --parallelism.expert_tensor_parallel_degree 2
```

I am testing it by doing an allgather + torch.equal check of the gate weight on the forward of `TokenChoiceTopKRouter` right before the gate forward is called, and indeed it fails after the first training step.

**Comment by volcacius** [comment]
My understanding of the issue is that the `NoParallel()` for the gate is embedded in a region with implicitly (as, non DTensor) sharded activation gradients since the expert weights are implicitly sharded (`to_local()` is called before using them with both EP and TP). I guess you could fix it by doing a reduction already on the re-ordering, but I think the issue is gonna show up if the re-ordering is not there. This is why I suggested fixing it by changing the `grad_placements` of `to_local` at the output of `NoParallel()`. Reordering is stateless anyway so I don't think the fix belongs there.

**Comment by tianyu-l** [comment]
> My understanding of the issue is that the NoParallel() for the gate is embedded in a region with implicitly (as, non DTensor) sharded activation gradients since the expert weights are implicitly sharded (to_local() is called before using them with both EP and TP).

This is different from my understanding.

My understanding of what happens on TP mesh with ETP = 1 is that
- router gate computation is entirely replicate -- replicate param, replicate input, replicate output
- reorderer will shard replicate input into sharded output -- this is where gradient flows back from sharded to full shape, but we forgot to do a reduction on the values so you would see discrepancy on what router backward sees

> TP=2,EP=2,ETP=2
With ETP=TP, my understanding has been that there is no numerics issue
- router computation is completely replicate in both fwd and bwd
- this line https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/expert_parallel.py#L211 will help mark gradient as Partial and reduce it to Replicate

I could be wrong. cc @acisseJZhong to follow up.

**Comment by acisseJZhong** [comment]
updated my findings in https://github.com/pytorch/torchtitan/pull/2416, there's numerics bug for both `ETP=1`(when using `ReordererSequenceParallel`) and `ETP=TP`(only when `score_before_experts=False`). The former is fixed by all reduce in `ReordererSequenceParallel` and the latter is fixed by all reduce when `score_before_experts=False`.

**Comment by acisseJZhong** [comment]
Pytorch side [PR](https://github.com/pytorch/pytorch/pull/175867/changes), torchtitan https://github.com/pytorch/torchtitan/pull/2416, please help take a review
